src/EditorFeatures/Core.Wpf/InlineParameterNameHints/InlineParameterNameHintsTagger.cs

```diff
@@ -19,56 +23,95 @@ internal sealed class InlineParameterNameHintsTagger : ITagger<IntraTextAdornmen
     {
         private readonly ITagAggregator<InlineParameterNameHintDataTag> _tagAggregator;
         private readonly ITextBuffer _buffer;
+        private readonly ITextView _textView;
+        private readonly List<ITagSpan<IntraTextAdornmentTag>> _cache;
+        private ITextSnapshot _cacheSnapshot;
+        private readonly IClassificationFormatMap _formatMap;
+        private TextFormattingRunProperties _format;
+        private readonly IClassificationType _keyword;
 
         public event EventHandler<SnapshotSpanEventArgs> TagsChanged;
 
-        public InlineParameterNameHintsTagger(ITextBuffer buffer, ITagAggregator<InlineParameterNameHintDataTag> tagAggregator)
+        public InlineParameterNameHintsTagger(InlineParameterNameHintsTaggerProvider taggerProvider, ITextView textView, ITextBuffer buffer, ITagAggregator<InlineParameterNameHintDataTag> tagAggregator)
         {
+            _cache = new List<ITagSpan<IntraTextAdornmentTag>>();
+            _textView = textView;
             _buffer = buffer;
             _tagAggregator = tagAggregator;
+            _formatMap = taggerProvider.ClassificationFormatMapService.GetClassificationFormatMap(textView);
+            _keyword = taggerProvider.ClassificationTypeRegistryService.GetClassificationType(InlineParameterNameHintsTag.TagId);
+            _formatMap.ClassificationFormatMappingChanged += this.OnClassificationFormatMappingChanged;
             _tagAggregator.TagsChanged += OnTagAggregatorTagsChanged;
         }
 
+        private void OnClassificationFormatMappingChanged(object sender, EventArgs e)
+        {
+            if (_format != null)
+            {
+                _format = null;
+                _cache.Clear();
+            }
+        }
+
         private void OnTagAggregatorTagsChanged(object sender, TagsChangedEventArgs e)
         {
+            _cache.Clear();
             var spans = e.Span.GetSpans(_buffer);
             foreach (var span in spans)
             {
                 TagsChanged?.Invoke(this, new SnapshotSpanEventArgs(span));
             }
         }
 
+        private TextFormattingRunProperties Format
+        {
+            get
+            {
+                _format ??= _formatMap.GetTextProperties(_keyword);
+
+                return _format;
+            }
+        }
+
         public IEnumerable<ITagSpan<IntraTextAdornmentTag>> GetTags(NormalizedSnapshotSpanCollection spans)
         {
-            if (spans.Count <= 0)
+            if (spans.Count == 0)
             {
                 return Array.Empty<ITagSpan<IntraTextAdornmentTag>>();
             }
 
-            var tagsList = new List<ITagSpan<IntraTextAdornmentTag>>();
-            var dataTags = _tagAggregator.GetTags(spans);
             var snapshot = spans[0].Snapshot;
-            foreach (var tag in dataTags)
+            if (_cache.Count == 0 || snapshot != _cacheSnapshot)
```

Comment:
So if the file has no spans, then we'll keep requerying the aggregator over and over again? Any concerns about performance or cache impacts there? I'm not sure if the cache needs to better distinguish "empty" versus "not cached".

Reply:
That's a good point. I will try to figure out a way to distinguish the cases.